java/src/main/java/ai/rapids/cudf/MemoryCleaner.java

```diff
@@ -214,10 +214,24 @@ static void setDefaultGpu(int defaultGpuId) {
         if (defaultGpu >= 0) {
           Cuda.setDevice(defaultGpu);
         }
+
         for (CleanerWeakReference cwr : all) {
```

Comment:
Cleaner has an API `isClean`, you can call it to do some check before really cleaning in `CleanerWeakReference.clean` to avoid duplicate cleaning.

Reply:
This bug was caused by `leak` instead of duplicate cleaning.
The resources should be closed by `cleaner.clean(logErrorIfNotClean = false)` trigged by `withResource` clause before this leak checking:
```
// leak checking in shutdown hook:
for (CleanerWeakReference cwr : all) {
  // this invoke cleaner.clean(logErrorIfNotClean = true)
  // If resource is not closed before, close it and then prints error log.
  cwr.clean(); 
}
```